common/lib/dependabot/clients/bitbucket_server.rb

```diff
@@ -0,0 +1,236 @@
+# frozen_string_literal: true
+
+require "dependabot/shared_helpers"
+require "faraday"
+require "faraday_middleware"
+
+module Dependabot
+  module Clients
+    class BitbucketServer
+      class NotFound < StandardError; end
+      class Unauthorized < StandardError; end
+      class Forbidden < StandardError; end
+
+      ##########
+      # Client #
+      ##########
+
+      def self.for_source(source:, credentials:)
+        credential = credentials.
+                     find do |cred|
+          cred["type"] == "git_source" &&
+            cred["host"] == source.hostname
+        end
+
+        new(source, credential)
+      end
+
+      def initialize(source, credential)
+        @source = source
+        @credentials = credential
+      end
+
+      def fetch_commit(_repo_name, branch)
+        path = "commits?limit=1&start=0&until=#{branch}"
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values", 0, "id")
+      end
+
+      def fetch_commit_message(branch)
+        path = "commits?limit=1&start=0&until=#{branch}"
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values", 0, "message")
+      end
+
+      def fetch_recent_commits(branch = nil)
+        path = "commits?limit=100&start=0"
+        path += "&until=#{branch}" if branch
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values")
+      end
+
+      def create_commit(branch_name, message, files)
+        files.each do |file|
+          content = Faraday::UploadIO.new(
+            StringIO.new(file.content), "application/text", file.path
+          )
+          body = {
+            content: content,
+            branch: branch_name,
+            message: message,
+            sourceCommitId: fetch_file_commit(branch_name, file.path)
+          }
+          url = File.join(repo_path, "browse", file.path)
+          put_multipart(url, body)
+        end
+      end
+
+      def fetch_file_commit(branch_name, file_path)
+        file_path = file_path.start_with?("/") ? file_path[1..-1] : file_path
+        path = "commits?limit=1&start=0&until=#{branch_name}&path=#{file_path}"
+        url = File.join(repo_path, path)
+
+        response = get(url)
+        response.body.dig("values", 0, "id")
+      end
+
+      def fetch_default_branch(_repo_name)
+        url = File.join(repo_path, "/branches/default")
+        response = get(url)
+
+        response.body.fetch("displayId")
+      end
+
+      def fetch_branch_by_name(branch)
+        url = File.join(repo_path, "/branches?filterText=#{branch}")
+        response = get(url)
+
+        response.body.dig("values")&.
+          find { |b| b["displayId"] == branch }&.
+          dig("id")
+      end
+
+      def create_branch(branch_name, base_commit)
```

Comment:
This will need some more error handling - apparently branch names are limited to 100 characters, and this method can produce branches that cause this error:

```
RuntimeError: Unhandled Bitbucket error!
Status: 400
Body: {"errors"=>[{"context"=>nil, "message"=>"The ref name, once fully qualified, is limited to 100 characters.", "exceptionName"=>"com.atlassian.bitbucket.repository.InvalidRefNameException", "refName"=>"dependabot/bundler/src/webapp/spec/acceptance/develop/rspec-and-rspec_junit_formatter-and-rspec-retry-3.9.0"}]}
  /var/lib/gems/2.6.0/bundler/gems/dependabot-core-cffb9144ef23/common/lib/dependabot/clients/bitbucket_server.rb:224:in `handle_response'
  /var/lib/gems/2.6.0/bundler/gems/dependabot-core-cffb9144ef23/common/lib/dependabot/clients/bitbucket_server.rb:215:in `request'
  /var/lib/gems/2.6.0/bundler/gems/dependabot-core-cffb9144ef23/common/lib/dependabot/clients/bitbucket_server.rb:234:in `block (2 levels) in <class:BitbucketServer>'
  /var/lib/gems/2.6.0/bundler/gems/dependabot-core-cffb9144ef23/common/lib/dependabot/clients/bitbucket_server.rb:104:in `create_branch'
```

I've worked around this temporarily by modifying the branch namer's `new_branch_name` method to return `branch_name.gsub("/", separator).chars.last(100).join`, but this isn't ideal.

Reply:
It turns out the qualified ref is prefixed with `refs/heads/`, so it's actually a max length of 110-11=89 characters. It also can't start with a `/` (stripping to the last 89 chars could cause this).

For now I'm getting by with

```ruby
module Dependabot
  class PullRequestCreator
    class BranchNamer
      old_new_branch_name = instance_method(:new_branch_name)
      define_method(:new_branch_name) do
        # The ref name in bitbucket API cannot be over 100 characters. It's prefixed with "refs/heads/".
        # It also cannot start with a "/".
        new_name = old_new_branch_name.bind(self).().chars.last(100 - "refs/heads/".length).join.gsub(/^\//,'')
      end
    end
  end
end
```

but it produces some odd branch names. Likely nobody will notice.